File: app/Domain/Messages/Message.php
<?php

namespace App\Domain\Messages {

    use App\Domain\Identity\UserId;
    use App\Domain\IDomainEvent;
    use App\Domain\IEventPublisher;
    use App\Domain\Messages\Message\DecisionProjection;

    class Message
    {
        public static function quack(IEventPublisher $eventPublisher, $messageContent, UserId $authorId)
        {
            $eventPublisher->publish(new MessageQuacked(MessageId::generate(), $messageContent, $authorId));
        }

        /**
         * Message constructor.
         * @param IDomainEvent[] $events
         */
        public function __construct(array $events)
        {
            $this->decisionProjection = new DecisionProjection($events);
        }

        public function requack(IEventPublisher $eventPublisher, UserId $requackerId)
        {
            if ($requackerId == $this->decisionProjection->getAuthorId()
                || in_array($requackerId, $this->decisionProjection->getRequackers())) {
                return;
            }
            $eventPublisher->publish(
                new MessageRequacked($this->decisionProjection->getMessageId(), $requackerId));
        }

        public function delete(IEventPublisher $fakeEventPublisher, UserId $anAuthorId): void
        {
            if ($this->decisionProjection->isDeleted()) {
                return;
            }

            $authorId = $this->decisionProjection->getAuthorId();
            if (!$authorId->equals($anAuthorId)) {
                return;
            }

            $fakeEventPublisher->publish(
                new MessageDeleted(
                    $this->decisionProjection->getMessageId(),
                    $anAuthorId
                )
            );
        }
    }
}

namespace App\Domain\Messages\Message {

    use App\Domain\DecisionProjectionBase;
    use App\Domain\Identity\UserId;
    use App\Domain\Messages\MessageDeleted;
    use App\Domain\Messages\MessageId;
    use App\Domain\Messages\MessageQuacked;
    use App\Domain\Messages\MessageRequacked;

    class DecisionProjection extends DecisionProjectionBase
    {
        /** @var MessageId */
        private $messageId;

        /** @var UserId */
        private $authorId;

        private $isDeleted = false;

        /** @var array */
        private $requackers = array();

        public function __construct($events)
        {
            $this->registerMessageQuacked();
            $this->registerMessageRequacked();
            $this->registerMessageDeleted();

            parent::__construct($events);
        }

        /**
         * @return MessageId
         */
        public function getMessageId()
        {
            return $this->messageId;
        }

        /**
         * @return UserId
         */
        public function getAuthorId()
        {
            return $this->authorId;
        }

        /**
         * @return array
         */
        public function getRequackers()
        {
            return $this->requackers;
        }

        private function registerMessageQuacked()
        {
            $this->register('App\Domain\Messages\MessageQuacked', function (MessageQuacked $event) {
                $this->messageId = $event->getMessageId();
                $this->authorId = $event->getAuthorId();
            });
        }

        private function registerMessageRequacked()
        {
            $this->register('App\Domain\Messages\MessageRequacked', function (MessageRequacked $event) {
                $this->requackers[] = $event->getRequackerId();
            });
        }

        private function registerMessageDeleted(): void
        {
            $this->register(MessageDeleted::class, function () {
                $this->isDeleted = true;
            });
        }

        public function isDeleted(): bool
        {
            return $this->isDeleted;
        }
    }
}

File: domain/src/main/scala/mixter/domain/message/Message.scala
package mixter.domain.message

import mixter.domain.identity.UserId
import mixter.domain.message.Message.DecisionProjection
import mixter.domain.message.event.{MessageDeleted, MessageEvent, MessageQuacked, MessageRequacked}
import mixter.domain.{Aggregate, EventPublisher}

case class Message(messageQuacked: MessageQuacked, events:Traversable[MessageEvent]) extends Aggregate {


  override type Id = MessageId
  override type AggregateEvent = MessageEvent
  override type InitialEvent = MessageQuacked

  private val projection = events.foldLeft(DecisionProjection.of(messageQuacked))((acc,event)=> acc(event))

  def requack(requacker: UserId, authorId:UserId, message:String)(implicit ep:EventPublisher): Unit =
    if(!projection.publishers.contains(requacker)){
      ep.publish(MessageRequacked(projection.messageId,requacker, authorId, message))
    }

  def delete(userId: UserId)(implicit ep:EventPublisher):Unit =
    if(projection.authorId==userId){
      ep.publish(MessageDeleted(projection.messageId))
    }
}

object Message {
  def quack(message: String, author: UserId)
           (implicit idGen:()=>MessageId=MessageId.generate, ep:EventPublisher)
  : Unit = {
    ep.publish(MessageQuacked(idGen(),message, author))
  }
  case class DecisionProjection(messageId:MessageId, authorId:UserId, publishers:Set[UserId]){
    def apply(messageEvent:MessageEvent):DecisionProjection = messageEvent match {
      case MessageRequacked(_,requacker, _, _)=> copy(publishers=publishers+requacker)
      case MessageDeleted(_) => this // effectless for now
      case MessageQuacked(_,_,_) => this // invalid, a message can only be requacked once
    }
  }
  object DecisionProjection{
    def of(messageQuacked: MessageQuacked) = DecisionProjection(messageQuacked.id, messageQuacked.author, Set(messageQuacked.author))
  }
}
